source/GameAction.cpp

```diff
@@ -0,0 +1,228 @@
+		GameData::Missions().Get(toFail);
+	}
+	else
+		conditions.Add(child);
+}
+
+
+
+void GameAction::Save(DataWriter &out) const
+{
+	if(!logText.empty())
+	{
+		out.Write("log");
+		out.BeginChild();
+		{
+			// Break the text up into paragraphs.
+			for(const string &line : Format::Split(logText, "\n\t"))
+				out.Write(line);
+		}
+		out.EndChild();
+	}
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+		{
+			out.Write("log", it.first, eit.first);
+			out.BeginChild();
+			{
+				// Break the text up into paragraphs.
+				for(const string &line : Format::Split(eit.second, "\n\t"))
+					out.Write(line);
+			}
+			out.EndChild();
+		}
+	if(payment)
+		out.Write("payment", payment);
+	for(const auto &it : events)
+	{
+		if(it.second.first == it.second.second)
+			out.Write("event", it.first->Name(), it.second.first);
+		else
+			out.Write("event", it.first->Name(), it.second.first, it.second.second);
+	}
+	for(const auto &name : fail)
+		out.Write("fail", name);
+	
+	conditions.Save(out);
+}
+
+
+
+bool GameAction::IsEmpty() const
+{
+	return empty;
+}
+
+
+
+int GameAction::Payment() const
+{
+	return payment;
+}
+
+
+
+// Do the actions of the GameAction. If this GameAction is a part of a conversations,
+// then some behavior may be different than from a MissionAction.
+void GameAction::Do(PlayerInfo &player, bool conversation) const
+{
+	if(!logText.empty())
+		player.AddLogEntry(logText);
+	for(const auto &it : specialLogText)
+		for(const auto &eit : it.second)
+			player.AddSpecialLog(it.first, eit.first, eit.second);
+	
+	if(payment)
+	{
+		// Conversation actions don't block a mission from offering if a
+		// negative payment would drop the player's account balance below
+		// zero, so payment is handled differently within conversations.
+		if(conversation)
+		{
+			int64_t account = player.Accounts().Credits();
+			// If the payment is negative and the player doesn't have enough
+			// in their account, then the player's credits are reduced to 0.
+			if(account + payment >= 0)
+				player.Accounts().AddCredits(payment);
+			else if(account > 0)
+				player.Accounts().AddCredits(-account);
+		}
+		else
+			player.Accounts().AddCredits(payment);
+	}
+	
+	for(const auto &it : events)
+		player.AddEvent(*it.first, player.GetDate() + it.second.first);
+	
+	if(!fail.empty())
+	{
+		// If this action causes this or any other mission to fail, mark that
+		// mission as failed. It will not be removed from the player's mission
+		// list until it is safe to do so.
+		for(const Mission &mission : player.Missions())
+			if(fail.count(mission.Identifier()))
+				player.FailMission(mission);
+	}
+	
+	// Check if applying the conditions changes the player's reputations.
+	player.SetReputationConditions();
+	conditions.Apply(player.Conditions());
+	player.CheckReputationConditions();
+}
+
+
+
+GameAction GameAction::Instantiate(map<string, string> &subs, int jumps, int payload) const
```

Comment:
I would expect subs to be const here as well.

Reply:
Sorry, just noticed that subs is changed here due to payment being set. That makes the code a little bit more complex indeed.